File: src/ai_legal_aid/interfaces/session.py
"""
Session management protocols for the AI Legal Aid System.

This module defines the contract for session management, including
session lifecycle, conversation history, and privacy-compliant
data handling.
"""

from abc import abstractmethod
from typing import Protocol

from ai_legal_aid.types import SessionId, Session


class SessionManager(Protocol):
    """Protocol for managing user sessions and conversation state."""

    @abstractmethod
    async def create_session(self) -> SessionId:
        """
        Create a new user session.

        Returns:
            Unique session identifier
        """
        ...

    @abstractmethod
    async def get_session(self, session_id: SessionId) -> Session:
        """
        Retrieve session information.

        Args:
            session_id: Unique session identifier

        Returns:
            Session information

        Raises:
            SessionNotFoundError: If session doesn't exist
        """
        ...

    @abstractmethod
    async def update_session(self, session_id: SessionId, updates: dict) -> None:
        """
        Update session information.

        Args:
            session_id: Unique session identifier
            updates: Dictionary of fields to update

        Raises:
            SessionNotFoundError: If session doesn't exist
        """
        ...

    @abstractmethod
    async def end_session(self, session_id: SessionId) -> None:
        """
        End a session and perform cleanup.

        Args:
            session_id: Unique session identifier

        Raises:
            SessionNotFoundError: If session doesn't exist
        """
        ...

    @abstractmethod
    async def cleanup_expired_sessions(self) -> None:
        """
        Clean up expired sessions and associated data.

        This method should be called periodically to maintain
        privacy compliance and system performance.
        """
        ...

File: src/interfaces/session.ts
/**
 * Session Management Interface
 * Handles user sessions, conversation state, and data privacy
 */

import { SessionId, Session } from '../types';

/**
 * Session Manager Interface
 * Manages user sessions with privacy-compliant data handling
 */
export interface SessionManager {
  /**
   * Create a new user session
   * @returns Promise resolving to new session ID
   */
  createSession(): Promise<SessionId>;

  /**
   * Retrieve an existing session by ID
   * @param sessionId - Unique session identifier
   * @returns Promise resolving to session data or null if not found
   */
  getSession(sessionId: SessionId): Promise<Session | null>;

  /**
   * Update session data with partial updates
   * @param sessionId - Unique session identifier
   * @param updates - Partial session data to update
   */
  updateSession(sessionId: SessionId, updates: Partial<Session>): Promise<void>;

  /**
   * End a session and perform cleanup
   * @param sessionId - Unique session identifier
   */
  endSession(sessionId: SessionId): Promise<void>;

  /**
   * Clean up expired sessions (privacy compliance)
   * @returns Promise resolving to number of sessions cleaned up
   */
  cleanupExpiredSessions(): Promise<number>;

  /**
   * Check if a session exists and is active
   * @param sessionId - Unique session identifier
   * @returns Promise resolving to boolean indicating session validity
   */
  isSessionActive(sessionId: SessionId): Promise<boolean>;

  /**
   * Get session statistics for monitoring
   * @returns Promise resolving to session statistics
   */
  getSessionStats(): Promise<{
    activeSessions: number;
    totalSessions: number;
    averageSessionDuration: number;
  }>;
}